### DISCUSSION TITLE
Completion in DAP REPL

### DISCUSSION THREAD

**Discussion by benjaminbauer** [LazyVim/LazyVim#3885] [Extra Requests]
I use the REPL in a debugger for trying things out. Having completion in the REPL (like VSCode does) would be great. The author of nvim-dap-ui provides a cmp source for this case https://github.com/rcarriga/cmp-dap

**Comment by Old-Farmer** [comment]
I don't know am I right, but this is my config and it seems work. I'm not confident of the correctness of this config so I will not make a pr. Anyone insterested can take a look.

```lua
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "rcarriga/cmp-dap",
    },
    opts = function(_, opts)
      local cmp = require("cmp")
      cmp.setup.filetype({ "dap-repl", "dapui_watches", "dapui_hover" }, {
        sources = {
          { name = "dap" },
        },
      })
      opts.enabled = function()
        local buftype = nil
        if vim.api.nvim_get_option_value then
          buftype = vim.api.nvim_get_option_value("buftype", { buf = 0 })
        else
          buftype = vim.api.nvim_buf_get_option(0, "buftype")
        end
        return buftype ~= "prompt" or require("cmp_dap").is_dap_buffer()
      end
    end,
  },
```

**Reply by LorenzoBettini** [reply]
By looking at the documentation https://github.com/rcarriga/cmp-dap, I have it working like this

```lua
{
  "hrsh7th/nvim-cmp",
  dependencies = {
    "rcarriga/cmp-dap",
  },
  opts = function()
    require("cmp").setup({
      enabled = function()
        return vim.api.nvim_buf_get_option(0, "buftype") ~= "prompt" or require("cmp_dap").is_dap_buffer()
      end,
    })

    require("cmp").setup.filetype({ "dap-repl", "dapui_watches", "dapui_hover" }, {
      sources = {
        { name = "dap" },
      },
    })
  end,
}
```

which I seem to understand is equivalent to yours.

**Reply by eagle-tex** [reply]
I looked at the documentation too and what you wrote seems correct, according to the documentation.
The problem is it isn't working in my config. Nothing happens, I still do not get the completions in the DAP-UI.

What does the `require("cmp")` stand for? It is referring to the nvim-cmp plugin or the cmp-dap one or another one?

**Reply by LorenzoBettini** [reply]
LazyVim has switched from cmp to blink by default, so the above will not work anymore unless you force cmp instead of blink.

**Reply by eagle-tex** [reply]
Thanks Lorenzo. I hope you don't mind if I ask you 2 questions.
1. How would you "force" it to use "cmp" instead of "blink"?
2. How would you change your code to use "blink" and provide code completions in DAP?

Please provide full workable snippets like the one you provided for the rcarriga plug-in, or a link to a repo if you have one ready.

Thanks in advance.

**Reply by LorenzoBettini** [reply]
> Thanks Lorenzo. I hope you don't mind if I ask you 2 questions.
> 
>     1. How would you "force" it to use "cmp" instead of "blink"?

See https://www.lazyvim.org/news

> to use nvim-cmp instead, enable the coding.nvim-cmp extra

> 
>     2. How would you change your code to use "blink" and provide code completions in DAP?

Still working on that
